agents/ten_packages/extension/openai_chatgpt_python/extension.py

```diff
@@ -0,0 +1,302 @@
+        if checking_vision_text_items_str:
+            try:
+                self.checking_vision_text_items = json.loads(checking_vision_text_items_str)
+            except Exception as err:
+                logger.info(f"Error parsing {PROPERTY_CHECKING_VISION_TEXT_ITEMS}: {err}")
+
+        # Create instance
+        try:
+            self.openai_chatgpt = OpenAIChatGPT(openai_chatgpt_config)
+            logger.info(f"initialized with max_tokens: {openai_chatgpt_config.max_tokens}, model: {openai_chatgpt_config.model}")
+        except Exception as err:
+            logger.info(f"Failed to initialize OpenAIChatGPT: {err}")
+
+        # Send greeting if available
+        if greeting:
+            try:
+                output_data = Data.create("text_data")
+                output_data.set_property_string(DATA_OUT_TEXT_DATA_PROPERTY_TEXT, greeting)
+                output_data.set_property_bool(DATA_OUT_TEXT_DATA_PROPERTY_TEXT_END_OF_SEGMENT, True)
+                ten_env.send_data(output_data)
+                logger.info(f"Greeting [{greeting}] sent")
+            except Exception as err:
+                logger.info(f"Failed to send greeting [{greeting}]: {err}")
+        ten_env.on_start_done()
+
+    def on_stop(self, ten_env: TenEnv) -> None:
+        logger.info("on_stop")
+
+        # TODO: clean up resources
+
+        ten_env.on_stop_done()
+
+    def on_deinit(self, ten_env: TenEnv) -> None:
+        logger.info("on_deinit")
+        ten_env.on_deinit_done()
+
+    def on_cmd(self, ten_env: TenEnv, cmd: Cmd) -> None:
+        logger.info(f"on_cmd json: {cmd.to_json()}")
+
+        cmd_name = cmd.get_name()
+
+        if cmd_name == CMD_IN_FLUSH:
+            self.outdate_ts = get_current_time()
+            ten_env.send_cmd(Cmd.create(CMD_OUT_FLUSH), None)
+            logger.info("on_cmd sent flush")
+            status_code, detail = StatusCode.OK, "success"
+        else:
+            logger.info(f"on_cmd unknown cmd: {cmd_name}")
+            status_code, detail = StatusCode.ERROR, "unknown cmd"
+        
+        cmd_result = CmdResult.create(status_code)
+        cmd_result.set_property_string("detail", detail)
+        ten_env.return_result(cmd_result, cmd)
+
+    def on_data(self, ten_env: TenEnv, data: Data) -> None:
+        # Get the necessary properties
+        is_final = get_property_bool(data, DATA_IN_TEXT_DATA_PROPERTY_IS_FINAL)
+        input_text = get_property_string(data, DATA_IN_TEXT_DATA_PROPERTY_TEXT)
+
+        if not is_final:
+            logger.info("ignore non-final input")
+            return
+        if not input_text:
+            logger.info("ignore empty text")
+            return
+
+        logger.info(f"OnData input text: [{input_text}]")
+
+        # Start an asynchronous task for handling chat completion
+        start_time = get_current_time()
+        asyncio.run_coroutine_threadsafe(self.__async_chat_completion(ten_env, start_time, input_text, self.memory), self.loop)
```

Comment:
additional wait previous task done maybe necessary, otherwise the `memory` maybe in wrong order. For instance, old task in `await` -> `flush` received -> new task start to run (without last memory) -> old task interrupted and append conversion in to memory.